File: Objective-C/AWScrollView-OC/AWScrollView.m
//
//  AWScrollView.m
//  AWScrollView-OC
//
//  Created by Alex Ling on 9/11/2015.
//  Copyright © 2015 Alex Ling. All rights reserved.
//

#import "AWScrollView.h"

@implementation AWScrollView

-(void)setUp{
	
	[self initialize];
	
	self.upPoint = CGPointMake(self.xExtension, 0);
	self.leftPoint = CGPointMake(0, self.yExtension);
	self.rightPoint = CGPointMake(2 * self.xExtension, self.yExtension);
	self.downPoint = CGPointMake(self.xExtension, 2 * self.yExtension);
	self.centerPoint = CGPointMake(self.xExtension, self.yExtension);
	
	self.frame = CGRectMake(0, 0, self.screenWidth, self.screenHeight);
	self.contentOffset = CGPointMake(self.xExtension, self.yExtension);
	self.contentSize = CGSizeMake(self.screenWidth + 2 * self.xExtension, self.screenHeight + 2 * self.yExtension);
	self.delegate = self;
	self.showsHorizontalScrollIndicator = NO;
	self.showsVerticalScrollIndicator = NO;
	self.directionalLockEnabled = YES;
	self.scrollEnabled = NO;
	self.bounces = NO;
	
	self.mainView = [[UIView alloc] initWithFrame:CGRectMake(self.xExtension, self.yExtension, self.screenWidth, self.screenHeight)];
	self.mainView.backgroundColor = [UIColor clearColor];
	
	UITapGestureRecognizer *tapRec = [[UITapGestureRecognizer alloc] initWithTarget:self action: @selector(mainViewTapped)];
	[self.mainView addGestureRecognizer:tapRec];
	
	[self addSubview:self.mainView];
	
	UIPanGestureRecognizer *panRec = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(paned:)];
	[self addGestureRecognizer:panRec];
}

-(void)mainViewTapped{
	[self scrollTo:CGPointMake(self.xExtension, self.yExtension)];
}

-(void)paned:(UIPanGestureRecognizer *)sender{
	if (sender.state == UIGestureRecognizerStateBegan){
		ScrollDirection scrollDir = None;
		CGPoint translation = [sender translationInView:self];
		if (fabs(translation.x) > fabs(translation.y)){
			if (translation.x > 0) {
				scrollDir = Left;
			}
			else{
				scrollDir = Right;
			}
		}
		else if (fabs(translation.x) < fabs(translation.y)){
			if (translation.y > 0) {
				scrollDir = Up;
			}
			else{
				scrollDir = Down;
			}
		}
		if (scrollDir != None) {
			[self handleScrollDirection:scrollDir];
		}
	}
}

-(void)initialize{
	if (!self.xExtension){
		self.xExtension = [UIScreen mainScreen].bounds.size.width/2;
	}
	
	if (!self.yExtension){
		self.yExtension = [UIScreen mainScreen].bounds.size.height/2;
	}
	
	if (!self.screenWidth){
		self.screenWidth = [UIScreen mainScreen].bounds.size.width;
	}
	
	if (!self.screenHeight){
		self.screenHeight = [UIScreen mainScreen].bounds.size.height;
	}
	
	if (!self.transitionTime){
		self.transitionTime = 0.5;
	}
}

-(void)handleScrollDirection:(ScrollDirection)direction{
	CGPoint offset = self.contentOffset;
	
	if (direction == Up) {
		if (self.contentOffset.x == self.xExtension){
			if (self.contentOffset.y == 2 * self.yExtension){
				offset = CGPointMake(self.xExtension, self.yExtension);
			}
			if (self.contentOffset.y == self.yExtension){
				offset = CGPointMake(self.xExtension, 0);
			}
		}
	}
	
	else if (direction == Down) {
		if (self.contentOffset.x == self.xExtension){
			if (self.contentOffset.y == 0){
				offset = CGPointMake(self.xExtension, self.yExtension);
			}
			if (self.contentOffset.y == self.yExtension){
				offset = CGPointMake(self.xExtension, 2 * self.yExtension);
			}
		}
	}
	
	else if (direction == Left) {
		if (self.contentOffset.y == self.yExtension){
			if (self.contentOffset.x == 2 * self.xExtension){
				offset = CGPointMake(self.xExtension, self.yExtension);
			}
			if (self.contentOffset.x == self.xExtension){
				offset = CGPointMake(0, self.yExtension);
			}
		}
	}
	
	else if (direction == Right) {
		if (self.contentOffset.y == self.yExtension){
			if (self.contentOffset.x == 0){
				offset = CGPointMake(self.xExtension, self.yExtension);
			}
			if (self.contentOffset.x == self.xExtension){
				offset = CGPointMake(2 * self.xExtension, self.yExtension);
			}
		}
	}
	
	[self scrollTo:offset];
}

-(void)scrollTo:(CGPoint)point{
	[UIView animateWithDuration:self.transitionTime animations:^{
		self.contentOffset = point;
	} completion: nil];
}

@end

File: AWScrollView/AWScrollView.swift
//
//  AWScrollView.swift
//  AWScrollView-Lab
//
//  Created by Alex Ling on 8/11/2015.
//  Copyright © 2015 Alex Ling. All rights reserved.
//

import UIKit

extension CGFloat{
	func closeTo (anotherFloat : CGFloat) -> Bool{
		return abs(self - anotherFloat) <= 0.5
	}
}

class AWScrollView: UIScrollView, UIScrollViewDelegate{
	
	let screenWidth = UIScreen.mainScreen().bounds.size.width
	let screenHeight = UIScreen.mainScreen().bounds.size.height
	
	var initialOffset : CGPoint = CGPoint()
	var locked : Bool = false
	
	var mainView : UIView!
	
	enum ScrollDirection {
		case Horizontal
		case Vertical
		
		case Up
		case Down
		
		case Left
		case Right
		
		case None
	}
	
	func setUp(){
		self.frame = CGRectMake(0, 0, self.screenWidth, self.screenHeight)
		self.contentOffset = CGPointMake(self.screenWidth/2, self.screenHeight/2)
		self.contentSize = CGSizeMake(2 * self.screenWidth, 2 * self.screenHeight)
		self.delegate = self
		self.showsHorizontalScrollIndicator = false
		self.showsVerticalScrollIndicator = false
		self.directionalLockEnabled = true
		self.bounces = false
		
		self.locked = true
		
		self.mainView = UIView(frame: CGRectMake(self.screenWidth/2, self.screenHeight/2, self.screenWidth, self.screenHeight))
		self.mainView.backgroundColor = UIColor.clearColor()
		
		let tapRec = UITapGestureRecognizer(target: self, action: Selector("mainViewTapped"))
		self.mainView.addGestureRecognizer(tapRec)
		
		self.addSubview(self.mainView)
	}
	
	func mainViewTapped(){
		self.scrollTo(CGPointMake(self.screenWidth/2, self.screenHeight/2))
	}
	
	func scrollViewWillBeginDragging(scrollView: UIScrollView) {
		self.initialOffset = self.contentOffset
	}
	
	func scrollViewDidScroll(scrollView: UIScrollView) {
		if (self.locked){
			let direction : ScrollDirection = self.determineScrollDirection(scrollView)
			var offset = self.initialOffset
			
			if direction == .Up {
				if self.initialOffset.x.closeTo(self.screenWidth/2) {
					if self.initialOffset.y.closeTo(self.screenHeight) || self.initialOffset.y.closeTo(self.screenHeight/2) {
						offset = CGPointMake(self.screenWidth/2, self.initialOffset.y - self.screenHeight/2)
					}
				}
			}
				
			else if direction == .Down {
				if self.initialOffset.x.closeTo(self.screenWidth/2) {
					if self.initialOffset.y.closeTo(0) || self.initialOffset.y.closeTo(self.screenHeight/2) {
						offset = CGPointMake(self.screenWidth/2, self.initialOffset.y + self.screenHeight/2)
						
					}
				}
			}
				
			else if direction == .Left {
				if self.initialOffset.y.closeTo(self.screenHeight/2) {
					if self.initialOffset.x.closeTo(self.screenWidth) || self.initialOffset.x.closeTo(self.screenWidth/2) {
						offset = CGPointMake(self.initialOffset.x - self.screenWidth/2, self.screenHeight/2)
					}
				}
			}
				
			else if direction == .Right {
				if self.initialOffset.y.closeTo(self.screenHeight/2) {
					if self.initialOffset.x.closeTo(0) || self.initialOffset.y.closeTo(self.screenHeight/2) {
						offset = CGPointMake(self.initialOffset.x + self.screenWidth/2, self.screenHeight/2)
					}
				}
			}
			self.scrollTo(offset)
		}
	}
	
	func determineScrollDirection(localSrcollView : UIScrollView) -> ScrollDirection {
		var scrollDir : ScrollDirection
		if self.initialOffset.x != localSrcollView.contentOffset.x && self.initialOffset.y != localSrcollView.contentOffset.y {
			if abs(self.initialOffset.x - localSrcollView.contentOffset.x) > abs(self.initialOffset.y - localSrcollView.contentOffset.y){
				scrollDir = .Horizontal
			}
			else{
				scrollDir = .Vertical
			}
		}
		else{
			if self.initialOffset.x == localSrcollView.contentOffset.x {
				scrollDir = .Vertical
			}
			else{
				scrollDir = .Horizontal
			}
		}
		if scrollDir == .Horizontal{
			if self.initialOffset.x > localSrcollView.contentOffset.x {
				scrollDir = .Left
			}
			else if self.initialOffset.x < localSrcollView.contentOffset.x {
				scrollDir = .Right
			}
			else {
				scrollDir = .None
			}
		}
		else{
			if self.initialOffset.y > localSrcollView.contentOffset.y {
				scrollDir = .Up
			}
			else if self.initialOffset.y < localSrcollView.contentOffset.y{
				scrollDir = .Down
			}
			else{
				scrollDir = .None
			}
		}
		return scrollDir
	}
	
	func scrollTo(point : CGPoint){
		self.locked = false
		self.scrollEnabled = false
		UIView.animateWithDuration(0.5, animations: {
			self.contentOffset = point
			}, completion: {(finished) in
				self.locked = true
				self.scrollEnabled = true
		})
	}
}

